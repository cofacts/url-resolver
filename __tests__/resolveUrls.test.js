const { gql } = require('../test-util');
const { closeBrowser } = require('../lib/scrap');

describe('resolveUrls', () => {
  it('resolves normal URLs', async () => {
    const result = await gql`
      {
        resolvedUrls(urls: ["https://example.com"]) {
          url
          canonical
          title
          summary
          html
          topImageUrl
          status
          error
        }
      }
    `();

    expect(result).toMatchSnapshot();
  });

  it('returns fetch error properly', async () => {
    const result = await gql`
      {
        resolvedUrls(
          urls: [
            "https://this-cannot-be-resolved.com"
            "line://ch/1341209850"
            "https://identityredesign.tw/vote-list.html" # has domain, but don't respond
            "http://beingsweetlife.com/archives/04/220087" # socket hang up without response (ERR_EMPTY_RESPONSE)
            "https://www.ey.gov.tw/File/66E9E54960EB958B?A=C" # PDF file, which is not supported
          ]
        ) {
          url
          error
        }
      }
    `();

    expect(result).toMatchSnapshot();
  });

  it.skip('resolves alert boxes', async () => {});

  it.skip('resolves redirects', async () => {});

  it('handles youtube URL', async () => {
    const result = await gql`
      {
        resolvedUrls(
          urls: [
            "https://www.youtube.com/watch?v=jNQXAC9IVRw"
            "https://www.youtube.com/watch?v=not-exist"
          ]
        ) {
          url
          canonical
          title
          summary
          status
          error
        }
      }
    `();

    expect(result).toMatchSnapshot();
  });

  it('returns minimal data on error', async () => {
    const result = await gql`
      {
        resolvedUrls(
          urls: [
            "http://blog.udn.com/watercmd/1066441" # https://github.com/cofacts/url-resolver/issues/2
          ]
        ) {
          url
          canonical
          title
          summary
          topImageUrl
          error
        }
      }
    `();

    expect(result).toMatchSnapshot();
  });

  it('handles URLs without protocol', async () => {
    const result = await gql`
      {
        resolvedUrls(urls: ["example.com"]) {
          url
          canonical
          title
        }
      }
    `();

    // Note that url should be same as input.
    //
    expect(result).toMatchSnapshot();
  });
});

afterAll(async () => {
  await closeBrowser();

  // eslint-disable-next-line no-console
  console.log('browser closed.');
});
