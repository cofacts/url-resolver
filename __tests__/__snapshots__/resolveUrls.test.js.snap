// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveUrls handles URLs without protocol 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "canonical": "http://example.com/",
        "title": "Example Domain",
        "url": "example.com",
      },
    ],
  },
}
`;

exports[`resolveUrls handles youtube URL 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "canonical": "https://youtu.be/jNQXAC9IVRw",
        "error": null,
        "status": 200,
        "summary": "The first video on YouTube. Maybe it's time to go back to the zoo? The name of the music playing in the background is Darude - Sandstorm.",
        "title": "Me at the zoo",
        "url": "https://www.youtube.com/watch?v=jNQXAC9IVRw",
      },
      Object {
        "canonical": null,
        "error": null,
        "status": 404,
        "summary": null,
        "title": null,
        "url": "https://www.youtube.com/watch?v=not-exist",
      },
    ],
  },
}
`;

exports[`resolveUrls resolves Javascript / html redirects 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "canonical": "https://www.tstartel.com/CWS/allrates-limitedtallk.php",
        "title": "- 台灣之星TSTAR",
        "url": "https://www.tstartel.com/static/discount/20161020_rollover/index.htm?utm_source=line&utm_medium=pic&utm_content=line_0504.rollover",
      },
      Object {
        "canonical": "",
        "title": "宜家卡 - IKEA",
        "url": "https://ikeabc.ecrm.com.tw/ikeaoa/l/11166",
      },
      Object {
        "canonical": "https://www.gvm.com.tw/article.html?id=42308",
        "title": "小心，愈來愈多人跟台灣說再見| 遠見雜誌",
        "url": "https://www.gvm.com.tw/article_content_33230.html",
      },
    ],
  },
}
`;

exports[`resolveUrls resolves normal URLs 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "canonical": "https://example.com/",
        "error": null,
        "html": "<!DOCTYPE html><html><head>
    <title>Example Domain</title>

    <meta charset=\\"utf-8\\">
    <meta http-equiv=\\"Content-type\\" content=\\"text/html; charset=utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <style type=\\"text/css\\">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: \\"Open Sans\\", \\"Helvetica Neue\\", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 50px;
        background-color: #fff;
        border-radius: 1em;
    }
    a:link, a:visited {
        color: #38488f;
        text-decoration: none;
    }
    @media (max-width: 700px) {
        body {
            background-color: #fff;
        }
        div {
            width: auto;
            margin: 0 auto;
            border-radius: 0;
            padding: 1em;
        }
    }
    </style>    
</head>

<body>
<div>
    <h1>Example Domain</h1>
    <p>This domain is established to be used for illustrative examples in documents. You may use this
    domain in examples without prior coordination or asking for permission.</p>
    <p><a href=\\"http://www.iana.org/domains/example\\">More information...</a></p>
</div>


</body></html>",
        "status": 200,
        "summary": "This domain is established to be used for illustrative examples in documents. You may use this
    domain in examples without prior coordination or asking for permission.
    More information...",
        "title": "Example Domain",
        "topImageUrl": "",
        "url": "https://example.com",
      },
    ],
  },
}
`;

exports[`resolveUrls returns fetch error properly 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "error": "NAME_NOT_RESOLVED",
        "url": "https://this-cannot-be-resolved.com",
      },
      Object {
        "error": "UNSUPPORTED",
        "url": "line://ch/1341209850",
      },
      Object {
        "error": "NOT_REACHABLE",
        "url": "https://identityredesign.tw/vote-list.html",
      },
      Object {
        "error": "NOT_REACHABLE",
        "url": "http://beingsweetlife.com/archives/04/220087",
      },
      Object {
        "error": "UNSUPPORTED",
        "url": "https://www.ey.gov.tw/File/66E9E54960EB958B?A=C",
      },
    ],
  },
}
`;

exports[`resolveUrls returns minimal data on error 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "canonical": "http://blog.udn.com/watercmd/1066441",
        "error": null,
        "summary": "主題：『吃香腸配養樂多會致癌嗎？』 我也是台客    轉貼: 　　導正『吃香腸配養樂多會致癌嗎？』的錯誤觀念　陳柏瑞 　　硝酸鹽和亞硝酸鹽是廣泛存在於自然環境中的化學物質，特別是在食物中，如蔬菜、 肉...",
        "title": "主題：吃香腸配養樂多會致癌嗎？ - 查無此人 - udn部落格",
        "topImageUrl": "https://g.udn.com.tw/community/img/common_all/blog_ogimg.jpg",
        "url": "http://blog.udn.com/watercmd/1066441",
      },
      Object {
        "canonical": "https://pension.president.gov.tw/cp.aspx?n=0710ED8C9356A871",
        "error": null,
        "summary": "網路流傳
			
			
			公教年金為何債台高築，據考試院不願具名官員私下表示，84年修法時巳刪除公教人員84年以後的年資退休後享有18%優惠存款的規定。卻在93年1月1日民進黨訂頒<政務人員退職撫卹條例>。就如蔡英文總統政大教授年資13年(73-79,82-89)，政務官5年(陸委會89-93,副院長95-97)，合計年資18年5個月，未達公教人員最低50歲，任職滿25年得退休的規定。由於民進黨政府93年訂頒<政務人員退職撫卹條例>, 蔡總統97年得以43歲副院長，年資18年5個月辦理退休，因退休時副院長薪資相當高，可辦18％優存金額410萬8600元全存入戶頭，每月的利息才62,829元。而93年訂頒的<政務人員退職撫卹條例>，規定可溯及既往，難怪當年起公保年金支出暴增，且持續增加，本來很多公務員要吃的這鍋公保年金就相當有限，突然跑來一堆大胃王（政務官）來搶食，公保年金不被吃跨才怪，由於這些人享有巨額退休金，就如馬前總統提到的，因而引起連戰和李元簇等不高興。

			民進黨的高官大口大口吞公保年金，如姚嘉文、游錫堃、吳榮義、郭瑤琪、杜正勝、陳哲男、朱武獻…..，口口聲聲罵18％，卻大手拿18％，真是無恥到極點。蘇貞昌也一再反18％，自稱沒有去領，問題錢入到他的戶頭就是他的，請問終止了該帳戶嗎？姚嘉文辯解說他雖享有18%優利存款，但他主張改革有什麼不對？公保年金被這些政務官大把大把搬光了，再來要求其他小公務員縮衣節食，硬要人多繳費少領錢，天底下就是有這麼不公平的事。
			
			
			這篇網路流傳的文章謬誤至為明顯如下：

			
				
				「政務人員退職酬勞金給與條例」於民國61年2月5日制定之初已有任滿2年可併計事務官年資領取一次退職金規定。至民國74年11月29日修正第四條又增列可領 月退職酬勞金規定。全文在民國88年6月15日修正，計算退職金基礎修正，且還回溯至民國85年5月1日開始適用。根本與93年施行的「政務人員退職撫卹條例」無關。
				
				
				而一次退職酬勞金存入18%優惠存款，更是民國61年10月26日就已訂定的「退職政務人員一次退職酬勞金優惠存款辦法」所規定。也不是93年施行的「政務人員退職撫卹條例」所訂定。政務官任滿2年、年資併計、退職酬勞金存18%都是在民國88年以前就存在的架構，已適用多年，且於93年後就已廢止18%優惠存款，改為一次離職金了。
				
				
				亦即，我國政務人員退職制度在92年以前，是採行「確定給付制」，自93年1月1日「政務人員退職撫卹條例」施行後，已改為「確定提撥之離職儲金制」。依現行規定，93年1月1日以後由軍、公、教人員、其他公職人員或公營事業人員轉任政務人員者，在職期間按月提撥離職儲金，退職時領取公、自提儲金本息（非由軍、公、教人員、其他公職人員或公營事業人員轉任政務人員者，無法參加政務人員離職儲金）；因此，93年以後擔任政務人員者，已無法領取退職酬勞金。目前只有少部分於92年以前已任政務人員且符合一定條件者，退職時仍可適用92年以前原規定請領退職酬勞金及辦理優惠存款。所以並無網路流傳文章所稱「政務人員退職撫卹條例規定讓政務人員享有巨額退休金」的情事。
				
				
				另外，就算是適用92年以前原規定請領退職酬勞金及辦理優惠存款的政務人員，其優惠存款金額也一體適用銓敘部於95年2月16日、100年1月1日及同年2月1日推動的調整方案進行調降，其中95年2月16日實施的方案對於支領月退職酬勞金的政務人員退休所得替代率上限訂為不超過現職待遇的75%至80%（較一般基層公務人員退休所得替代率上限為85%至95%更為嚴格），超過者必須調降養老給付優惠存款金額。至於100年2月1日實施的現行方案已明定支領月退職酬勞金的退職特任政務人員公保養老給付優惠存款金額不得超過200萬元。
				
			

			以上說明，自95年以來實施的歷次優惠存款調整方案，都一體適用於政務人員。這些因制度過渡期產生的人數不多，所占比率也小，不可能有如網路流傳文章所稱「公保年金被政務人員大把搬光，要求其他小公務員縮衣節食」的情事；更不應該成為年金制度不需改革的藉口。",
        "title": "輿情澄清4：政務人員退職金吃垮公保年金嗎?",
        "topImageUrl": "",
        "url": "https://pension.president.gov.tw/cp.aspx?n=0710ED8C9356A871",
      },
      Object {
        "canonical": "https://www.fun123.com.tw/site/promote/download_app",
        "error": null,
        "summary": "流行3C周邊商品 限時特價",
        "title": "3C市集 | 流行3C周邊商品 限時特價",
        "topImageUrl": "https://www.fun123.com.tw/images/desktop/downloadapp/fun123_downloadapp_banner.jpg",
        "url": "http://ms7.tw/DL/D?k=app_daily_coupon",
      },
      Object {
        "canonical": null,
        "error": "NOT_REACHABLE",
        "summary": null,
        "title": null,
        "topImageUrl": null,
        "url": "https://99md.cn/fslh",
      },
    ],
  },
}
`;
